### ISSUE TITLE
FR: Add a command for signing commits (when `signing.sign-all = false`)

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#4712]
**Is your feature request related to a problem? Please describe.**

We have had support for automatically signing all commits since #3007.

**Describe the solution you'd like**

We want a `jj sign` command for when `signing.sign-all = false`.

**Describe alternatives you've considered**

None.

**Additional context**

This ticket was extracted from #58 so we could close that one and not make people think that we don't have any support for signing at all.

**Comment by pylbrecht** [comment]
I am going to take a stab at this (thanks again, @martinvonz, for the recommendation).

My usual progress disclaimer: family + full time job = 🐌

**Comment by yuja** [comment]
Maybe you can extract some bits from https://github.com/martinvonz/jj/pull/3142

**Comment by pylbrecht** [comment]
Oh, I wasn't aware that there is already an open PR for this.

@julienvincent, do you want to continue working on https://github.com/martinvonz/jj/pull/3142?
I am happy to pick another issue instead.

**Comment by julienvincent** [comment]
@pylbrecht Please go ahead - I was planning on coming back to this at some point but I don't think I'm going to get a chance any time soon.

**Comment by martinvonz** [comment]
> Oh, I wasn't aware that there is already an open PR for this.

Sorry, I had forgotten that we had this pending PR. Thanks for pointing it you, @yuja.

**Comment by pylbrecht** [comment]
Heartbeat: I am still digging around the linked issues and PRs to build up context.

It looks like `jj sign` needs to be built on top of #3142, as it implements displaying commit signatures. Without it we do not have anything in stdout to assert against in CLI tests.

@julienvincent, I suppose you will not be able to work on #3142 any time soon either?

**Comment by julienvincent** [comment]
I suppose you mean #3141? I might be able to pick it up again on a weekend, but I wouldn't be able to commit very hard to it right now.

If you want to pick up work I don't have any particular attachments to it :)

**Comment by pylbrecht** [comment]
Oh yeah, I meant #3141 (fixed the typo in the [original comment](https://github.com/martinvonz/jj/issues/4712#issuecomment-2467950785)).

> I might be able to pick it up again on a weekend, but I wouldn't be able to commit very hard to it right now.
>
> If you want to pick up work I don't have any particular attachments to it :)

Well, my first steps would be to take your changes, understand them, and eventually add whatever seems to be missing to start working on #4747. While there is a great chance it will take me as long (or even longer) as you to finish the job, 
I would still give it a try, for the sake of learning.

**Comment by ElvishJerricco** [comment]
I don't know if I should make a different issue for this or just leave a comment here, but I think the nicest workflow (for me, anyway) would be to sign commits automatically on push. i.e. I don't necessarily want to unlock my GPG key every time jj makes a snapshot, but I probably do want to sign anything that I push to a remote without having to manually run a sign command.

**Comment by emilazy** [comment]
Agreed, batch signing should be a pre‐publish thing, like other clean‐up/linting hooks. (We were just discussing this on the Discord today.)

**Comment by martinvonz** [comment]
I was going to say that we can do that in `report_repo_changes()` here: https://github.com/martinvonz/jj/blob/0de36918e4c020e0e54816f29c47cb57cc9cfbf5/cli/src/cli_util.rs#L2032-L2038

However, as the comment there says, there may be millions of new commits added to the repo, so it can be very expensive to check all of them. It's probably better to go through `MutableRepo::parent_mapping` to find the rewritten commits: https://github.com/martinvonz/jj/blob/0de36918e4c020e0e54816f29c47cb57cc9cfbf5/lib/src/repo.rs#L870

However, we currently clear that map after using it. I don't think we should have to clear. I pushed https://github.com/martinvonz/jj/compare/main...push-rmnzxqzrwxwp to show what I mean, but I'm not at all confident that that's actually safe. I'll need to think more about it later.

**Comment by bnjmnt4n** [comment]
> I don't know if I should make a different issue for this or just leave a comment here, but I think the nicest workflow (for me, anyway) would be to sign commits automatically on push. i.e. I don't necessarily want to unlock my GPG key every time jj makes a snapshot, but I probably do want to sign anything that I push to a remote without having to manually run a sign command.

I quite like this approach as well, and I've created a PR adding a `git.sign-on-push` configuration option which does this that I've been using locally to push commits without having to sign them all the time (https://github.com/jj-vcs/jj/pull/5164). I think this doesn't replace an explicit `jj sign` command, but could be an alternative/accompaniment. (It currently depends on some commits from #4853 to add some basic tests.)

**Comment by pylbrecht** [comment]
With #4853 merged, I am now focusing on getting #4747 out the door.

**Comment by pylbrecht** [comment]
It's finally done.
Thanks to everybody involved and especially thanks to @yuja for *so many*, very timely code reviews.

I created some follow-ups from discussions, which we still need to tackle:
* https://github.com/jj-vcs/jj/issues/5428
* https://github.com/jj-vcs/jj/issues/5786
* https://github.com/jj-vcs/jj/issues/5804
